File: app/src/main/java/com/itba/runningMate/repository/run/RunRepositoryImpl.java
package com.itba.runningMate.repository.run;

import com.itba.runningMate.db.RunDao;
import com.itba.runningMate.domain.Run;
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;

public class RunRepositoryImpl implements RunRepository {

    private final RunDao runDao;
    private final SchedulerProvider scheduler;

    public RunRepositoryImpl(final RunDao runDao, final SchedulerProvider scheduler) {
        this.runDao = runDao;
        this.scheduler = scheduler;
    }

    @Override
    public Flowable<List<Run>> getRun() {
        return runDao.getRoutes().map(RunMapper::toModel);
    }

    @Override
    public Flowable<List<Run>> getRunLazy() {
        return runDao.getRoutesLazy().map(RunMapper::toModel);
    }

    @Override
    public Single<Run> getRunMetrics(long uid) {
        return runDao.getRouteMetrics(uid).map(RunMapper::toModel);
    }

    @Override
    public Single<Run> getRun(long uid) {
        return runDao.getRoute(uid).map(RunMapper::toModel);
    }

    @Override
    public Single<Long> insertRun(Run run) {
        return runDao.insertRoute(RunMapper.toEntity(run));
    }

    @Override
    public Single<Double> getTotalDistance() {
        return runDao.getTotalDistance();
    }

    @Override
    public Single<Long> getMaxTime() {
        return runDao.getMaxTime();
    }

    @Override
    public Single<Double> getMaxKcal() {
        return runDao.getMaxKcal();
    }

    @Override
    public Single<Double> getMaxSpeed() {
        return runDao.getMaxSpeed();
    }

    @Override
    public Completable updateTitle(long runId, String title) {
        return runDao.updateTitle(runId, title);
    }

    @Override
    public void deleteRun(Run run) {
        runDao.deleteRoute(RunMapper.toEntity(run))
                .onErrorComplete()
                .subscribeOn(scheduler.io())
                .observeOn(scheduler.ui())
                .subscribe();
    }


    public void deleteRun(long runId) {
        runDao.deleteRoute(runId)
                .onErrorComplete()
                .subscribeOn(scheduler.io())
                .observeOn(scheduler.ui())
                .subscribe();
    }
}

File: app/src/main/java/com/itba/runningMate/repository/run/RunRepositoryImpl.kt
package com.itba.runningMate.repository.run

import com.itba.runningMate.db.run.RunDao
import com.itba.runningMate.db.run.RunEntity
import com.itba.runningMate.domain.Run
import com.itba.runningMate.repository.run.RunMapper.Companion.toEntity
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single

class RunRepositoryImpl(private val runDao: RunDao) : RunRepository {
    override fun getRun(): Flowable<List<Run>> {
        return runDao.getRoutes()
            .map { obj: List<RunEntity> -> RunMapper.toModel(obj) }
    }

    override fun getRunLazy(): Flowable<List<Run>> {
        return runDao.getRoutesLazy()
            .map { obj: List<RunEntity> -> RunMapper.toModel(obj) }
    }

    override fun getRunMetrics(uid: Long): Single<Run> {
        return runDao.getRouteMetrics(uid)
            .map { entity: RunEntity -> RunMapper.toModel(entity) }
    }

    override fun getRun(uid: Long): Single<Run> {
        return runDao.getRoute(uid)
            .map { entity: RunEntity -> RunMapper.toModel(entity) }
    }

    override fun insertRun(run: Run): Single<Long> {
        return runDao.insertRoute(toEntity(run))
    }

    override fun getTotalDistance(): Single<Double> {
        return runDao.getTotalDistance().switchIfEmpty(Single.just(0.0))
    }

    override fun getMaxTime(): Single<Long> {
        return runDao.getMaxTime().switchIfEmpty(Single.just(0))
    }

    override fun getMaxKcal(): Single<Double> {
        return runDao.getMaxKcal().switchIfEmpty(Single.just(0.0))
    }

    override fun getMaxSpeed(): Single<Double> {
        return runDao.getMaxSpeed().switchIfEmpty(Single.just(0.0))
    }

    override fun updateTitle(runId: Long, title: String): Completable {
        return runDao.updateTitle(runId, title)
    }

    override fun deleteRun(run: Run): Completable {
        return runDao.deleteRoute(toEntity(run))
    }

    override fun deleteRun(runId: Long): Completable {
        return runDao.deleteRoute(runId)
    }
}